sys/net/gnrc/netif/gnrc_netif.c

```diff
@@ -1305,63 +1337,93 @@ static void *_gnrc_netif_thread(void *args)
     gnrc_netif_release(netif);
 
     while (1) {
-        DEBUG("gnrc_netif: waiting for incoming messages\n");
-        msg_receive(&msg);
-        /* dispatch netdev, MAC and gnrc_netapi messages */
-        switch (msg.type) {
-            case NETDEV_MSG_TYPE_EVENT:
-                DEBUG("gnrc_netif: GNRC_NETDEV_MSG_TYPE_EVENT received\n");
-                dev->driver->isr(dev);
-                break;
-            case GNRC_NETAPI_MSG_TYPE_SND:
-                DEBUG("gnrc_netif: GNRC_NETDEV_MSG_TYPE_SND received\n");
-                res = netif->ops->send(netif, msg.content.ptr);
-                if (res < 0) {
-                    DEBUG("gnrc_netif: error sending packet %p (code: %u)\n",
-                          msg.content.ptr, res);
+        int msg_waiting = 0;
+#ifdef MODULE_GNRC_NETIF_EVENTS
+        /* Using messages for external IPC, and events for internal events */
+        DEBUG("gnrc_netif: waiting for events\n");
+        /* We can not use event_loop() or event_wait() because then we would not
+         * wake up when a message arrives */
+        thread_flags_t flags = thread_flags_wait_any(
+            THREAD_FLAG_MSG_WAITING | THREAD_FLAG_EVENT);
+        /* Events will be handled before messages */
+        if (flags & THREAD_FLAG_EVENT) {
+            DEBUG("gnrc_netif: handling events\n");
+            event_t *evp;
+            while ((evp = event_get(&netif->evq))) {
+                DEBUG("gnrc_netif: event %p\n", (void *)evp);
+                if (evp->handler) {
+                    evp->handler(evp);
                 }
-                break;
-            case GNRC_NETAPI_MSG_TYPE_SET:
-                opt = msg.content.ptr;
+            }
+        }
+        if (flags & THREAD_FLAG_MSG_WAITING) {
+            /* non-blocking msg check */
+            msg_waiting = msg_try_receive(&msg);
+        }
+#else /* MODULE_GNRC_NETIF_EVENTS */
+        /* Only messages used for event handling */
+        DEBUG("gnrc_netif: waiting for incoming messages\n");
+        msg_waiting = msg_receive(&msg);
+#endif /* MODULE_GNRC_NETIF_EVENTS */
```

Comment:
This whole `ifdef` block should IMHO go into its own function with `msg_waiting` as a return value.

Reply:
(this would also make the diff more readable ;-))

Reply:
Addressed